java/arcs/core/policy/PolicyConstraints.kt

```diff
@@ -164,4 +158,23 @@ sealed class PolicyViolation(val policy: Policy, message: String) : Exception(
         policy: Policy,
         checks: List<Check>
     ) : PolicyViolation(policy, "Recipe violates egress checks: $checks")
+
+    /** Thrown when a recipe is missing the `@policy` annotation. */
+    class MissingPolicyAnnotation(recipe: Recipe, policy: Policy) : PolicyViolation(
+        policy,
+        "Recipe '${recipe.name}' does not have an @policy annotation."
+    )
+
+    /** Thrown when a recipe was checked against a mismatched policy. */
+    class MismatchedPolicyName(electedPolicyName: String, policy: Policy) : PolicyViolation(
+        policy,
+        "Recipe elected a policy named '$electedPolicyName'."
+    )
+}
+
+/** Converts a list of particles into their names and egress types, as a string. */
+private fun List<ParticleSpec>.namesAndEgressTypes(): String {
```

Comment:
Could be a private function in`InvalidEgressTypeForParticles` unless you are reusing it?